### DISCUSSION TITLE
Target container is not a DOM element.

### DISCUSSION THREAD

**Discussion by jpriollaud** [ElMassimo/vite_ruby#290] [Q&A] [answered]
I am getting "Target container is not a DOM element." when trying to render React Portals in 2 child nodes in my `<div id=root>`

// application layout (not using a view because this is nav-bar and footer site-wide and I want to DRY)

**Scenario 1 (duplicated div tags)**
If I place `<div id=root>` duplicated in 2 places, in `<head>` and `<body>` then the Target container is found and everything works.

**Scenario 2 (no duplicate div tags)**
If I try to place <%= vite_javascript_tag 'foo.jsx' %> **after** my lone `<div id=root>` in my body somehow the script is still running prematurely before the root element is available.  Even if I try to tell React to wait for 'DOMContentLoaded' before doing anything, the entrypoint is still executing prematurely.  

I have `<%= vite_client_tag %>` in my `head`, along with a few other vite_javascript_tags for which load order does not matter.

I tried moving this React entrypoint out of entrypoints and into a subfolder, it did not help.

Is vite_ruby acceleration logic preventing me from dictating load order?

**Comment by ElMassimo** [comment, marked as answer]
Hi there!

> Is vite_ruby acceleration logic 

Not sure what you are referring to 😃 

---

By default all scripts in Vite are [`type=module`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#other_differences_between_modules_and_standard_scripts), which by are [`defer`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-defer) by default.

> Scripts with the defer attribute will execute in the order in which they appear in the document.

> ...the script is meant to be executed after the document has been parsed, but before firing [DOMContentLoaded](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event).

If an HTML element is defined (and closed) before a script element, it will __always__ be available when the script is executed (in fact that's exactly the technique [I'm using to hydrate islands in îles](https://iles.pages.dev/)).

The issue you are experiencing suggests malformed HTML, or that the script is inside the root element.

---

If you share a repo with a reproduction I can take a look. Regardless, there's nothing in Vite Ruby that would affect order of execution.

**Reply by jpriollaud** [reply]
That's helpful @ElMassimo -I should have noticed that module/defer.  (by acceleration I meant whatever vite_ruby is doing under the hood to speed up asset loading and rendering on bin/vite dev)

I'm pretty sure this is not a vite_ruby issue, and some weird React 17 behavior.  But here's what I have:

// Works, only because of duplicate elements.  Truly strange.  If I a remove a dupe I get 'not a DOM element' errors raised in both portals.

Rendered html:

```html
<div id="react">
  <div id="nav-bar"></div>
  <div id="signed-in-footer"></div>
</div>

<div id="react">
  <div id="nav-bar"></div>
  <div id="signed-in-footer"></div>
</div>

<script src="/vite-dev/entrypoints/navs.jsx" crossorigin="anonymous" type="module"></script>
```

navs.jsx:

```javascript
import React from 'react'
import ReactDOM from 'react-dom';

import ResponsiveAppBar from '~/components/app_bar'
import ResponsiveAppFooter from '~/components/app_footer'
import { UserSettingsProvider } from '~/components/theme_context'

const AppBarPortal = () => ReactDOM.createPortal(<ResponsiveAppBar />, document.getElementById('nav-bar'))
const AppFooterPortal = () => ReactDOM.createPortal(<ResponsiveAppFooter />, document.getElementById('signed-in-footer'))

ReactDOM.render(
  <React.StrictMode>
    <UserSettingsProvider>
      <AppBarPortal />
      <AppFooterPortal />
    </UserSettingsProvider>
  </React.StrictMode>,
  document.getElementById('react')
);


```

**Reply by ElMassimo** [reply]
Any DOM elements will be exposed as variables named after their `id` in the global object (`window`):

<img width="442" alt="Screen Shot 2022-10-27 at 21 35 59" src="https://user-images.githubusercontent.com/1158253/198422483-fef8a226-c879-46a9-819a-787c1c3d1492.png">

The fact that you are using `id="react"` could be somehow related to the "not a DOM element" error as it's not unusual to set `react` as a global.

Try using a more "unique" id for elements, such as `id="app"` or `id="app-root"`.

**Comment by jpriollaud** [comment]
no change unfortunately.

`<div id="root">` => Target container is not a DOM element.

`<div id="root">` `<div id="root">` => works.

This must be a strange bug in React.  I'll try to refactor.

**Comment by jpriollaud** [comment]
Solved:

```html
<div id="react"></div>
<div id="nav-bar"></div>
<div id="signed-in-footer"></div>
```

React doesn't like nested elements for some reason.  Sorry for the off topic @ElMassimo
